spec/value_inspector_spec.rb

```diff
@@ -61,9 +66,154 @@
     let(:value) { [{ id: 1, contents: "some text" }, { id: 2 }] }
     it "is an array containing the schema" do
       expect(schema).to eq(
+        class: "Array",
+        items: [
+          {
+            class: "Hash",
+            properties: [
+              { name: :id, class: "Integer" },
+              { name: :contents, class: "String" },
+            ],
+          },
+          {
+            class: "Hash",
+            properties: [ { name: :id, class: "Integer" } ]
+          }
+        ],
+      )
+    end
+
+    describe "max depth" do
+      let(:max_depth) { 0 }
+      it "exceeds max_depth to describe element types when inspecting arrays" do
+        expect(schema).to eq(
+          class: "Array",
+          items: [ { class: "Hash" } ]
+        )
+      end
+    end
+  end
+
+  describe "Nested arrays" do
+    let(:value) { [ [[[1]], [[2]]], [[[3]], [[4]]] ] }
+    let(:max_depth) { 1 }
+    it "exceeds max depth to get type information" do
+      expect(schema).to eq(
+        class: "Array",
+        items: [
+          {
+            class: "Array",
+            items: [
+              class: "Array",
+              items: [
+                {
+                  class: "Array",
+                  items: [
+                    { class: "Integer" }
+                  ]
+                }
+              ]
+            ]
+          }
+        ]
+      )
+    end
+
+    describe "with Hashes" do
+      let(:max_depth) { 2 }
+      let(:value) do
+        [
+          {
+            id: 0,
+            children: [
+              { id: 1, children: [ { id: 2 } ] },
+              { id: 3, children: [ { id: 4 } ] }
+            ]
+          }
+        ]
+      end
+
+      it "won't exceed max_depth when inspecting a Hash" do
+        expect(schema).to eq(
+          class: "Array",
+          items: [
+            {
+              class: "Hash",
+              properties: [
+                { name: :id, class: "Integer" },
+                { name: :children, class: "Array", items: [ { class: "Hash" } ] }
+              ]
+            }
+          ]
+        )
+      end
+    end
+  end
+
+  describe "Array of Strings" do
+    let(:value) { %w[one two three] }
+    it "is an array containing the schema" do
+      expect(schema).to eq(
+        class: "Array",
+        items: [
+          { class: "String" },
+        ],
+      )
+    end
+  end
+
+  describe "Mixed array" do
+    let(:value) { [ 1, "two", { key: "three" } ] }
+    it "correctly describes the schema" do
+      expect(schema).to eq(
+        class: "Array",
+        items: [
+          { class: "Integer" },
+          { class: "String" },
+          { class: "Hash", properties: [ { name: :key, class: "String" } ] },
+        ],
+      )
+    end
+  end
+
+  describe "Max array elements" do
+    let(:value) { [ 1, 'two', { id: 3 }, [ 'four' ] ] }
+    let(:max_array_elements) { 3 }
+
+    it "only describes the first N elements" do
```

Comment:
I think that the schema of the fourth element would be merged in with the schema of the second, due to the `uniq`, so I don't think this test actually verifies the truncation at N=4.

Reply:
This'll end up resolving to
```yml
class: Array
items:
  - class: Integer
  - class: String
  - class: Hash
    properties:
      - name: id
        class: Integer
  - class: Array
    items:
      - class: String
```

The `uniq` is using each items `hash` which should be unique in this case.